### ISSUE TITLE
Skimming through book bricks Kobo Forma.

### ISSUE THREAD

**Issue by chris-montero** [koreader/koreader#12633]
* KOReader version: 2024.07 "Outside"
* Device: Kobo Forma

#### Issue
Skimming through a book bricked my device. I had to do a full factory reset.

#### Steps to reproduce

Notes: 
* I used [this](https://github.com/koreader/koreader/wiki/Installation-on-Kobo-devices#alternative-manual-installation-method-based-on-nickelmenu) installation method, and followed it correctly.
* KOreader worked correctly on my device for a few weeks until this happened.

1. Using epub version of this book: https://vulkan-tutorial.com/ (which can be downloaded for free from the website)
2. Open the ebook.
3. open the menu up top
4. Choose "skim document"
5. press -10 to go back every 10 pages until the beginning of the book.
6. press a few more times -10, even though it's already the beginning of the book
7. close the menu
8. press sleep button
9. It won't sleep
10. From this point on, the device is bricked. I had to press power + next_page button for 20 seconds to do a factory reset.

**Comment by benoit-pierre** [comment]
Please provide some logs. You might want to try the latest nightly too. #11511 may be relevant.

**Comment by chris-montero** [comment]
@benoit-pierre I have no logs. The factory reset erased everything, including my books.

**Comment by benoit-pierre** [comment]
Force shutdown does not work?

**Comment by chris-montero** [comment]
I tried it, it did not work.

**Comment by benoit-pierre** [comment]
I don't have a kobo, but from a quick look around, it looks like holding longer trigger a force power off (but the screen does not change):

> Holding the power button down for five seconds or so should trigger a shutdown. When that happens it will display the powered off screen. If the device has hung, holding it down for longer than that, will force a power off. I'm not sure how long is needed, but, I always say 20 seconds in the factory reset instructions. When you do this, the screen doesn't change.

**Comment by chris-montero** [comment]
I kept holding the power off button for more than 1 minute and nothing happened.

**Comment by benoit-pierre** [comment]
FWIU, nothing happen then, the screen does not change, you have to press the power button again for ~3s to restart.

Anyway, without logs, it's going to be hard to diagnose. You could still give the nightly a go.

**Comment by hius07** [comment]
Cannot reproduce on Kindle.
Dunno if related, in the crash.log:
```
CRE: document loaded, but styles re-init needed (cause: peculiar CSS pseudoclasses met)
CRE: styles re-init needed after load, re-rendering

```

**Comment by poire-z** [comment]
No issue for me following the described "steps to reproduce" on my old Kobo Glo HD.
No problem with that book, it's not heavy, it opens fast, it turns pages fast. It goes to sleep and wakes up fine.

> CRE: document loaded, but styles re-init needed (cause: peculiar CSS pseudoclasses met)
> CRE: styles re-init needed after load, re-rendering

That's legitimate and normal, not an issue.

**Comment by chris-montero** [comment]
I installed everything again. Will ping back if it happens again.

**Comment by Frenzie** [comment]
> I kept holding the power off button for more than 1 minute and nothing happened.

I'm not sure otoh if it also works when you hold it for very long. Older devices were better in that they had a separate reset button, but the basic concept of holding the power button for 5-10 seconds, letting go for a few seconds, and pressing it again (maybe for a couple of seconds) should work well enough.

It's not really possible to brick the device without messing with startup scripts or similar.

**Comment by NiLuJe** [comment]
> It's not really possible to brick the device without messing with startup scripts or similar.

This.

----

That said, the Forma *is* on the list of "may randomly trigger EPDC hangs" devices.

But, like what was explained above, you recover by forcing a power-off (long [but not massively so, usually something north of 5s but south of 15s] press on power, usually clear enough when you had the frontlight on, because it will turn off), wait a few seconds, then power it back on (short press on power).

For it to trip a factory reset, either the storage was corrupted, or you *also* had a page-turn button pressed during boot, as *that* will force a factory reset. According to your post, you did boot that way: that was the mistake ;).

**Comment by chris-montero** [comment]
@NiLuJe You're wrong. I actually first tried to simply press the power button, hoping the tablet would wake up from sleep. When that didn't work, I tried to long-press the power button a few times for intervals of 5-60 seconds, hoping it would power off/on. That didn't do anything so my last resort was intentionally forcing a factory reset by pressing both next-page and the power button.
Hopefully this clarifies exactly the steps I went through before deciding to do a factory reset.
As I said, if I encounter something like this, I will reply to this issue, and I will try to get some crash logs if I can.

**Comment by NiLuJe** [comment]
Color me doubtful/surprised. I happen to own a day one Forma, so I'm intimately familiar with what happens when the driver decides to commit seppuku. Ever more so since mine started to do all kinds of extra weird shit irrelevant to the aforementioned EPDC hangs.

TL;DR: I've interacted with hung Formas *a lot* ;).

That's not to say that it wasn't a *wrong* reflex, I just wanted to clarify that that is by no means the first thing someone encountering a hung Kobo should attempt ;).

**Comment by NiLuJe** [comment]
> I will try to get some crash logs if I can

Unless you can catch it via serial, you likely won't, as these hangs do not happen while Wi-Fi is on, or while plugged it, which excludes monitoring over SSH either way.

And since the kernel hangs, nothing will get flushed to disk, so you're unlikely to see anything relevant anywhere persist on reboot.

(Again, lotsa experience here, I've tried all kinds of things to try to catch a trace of these, to no avail).

**Comment by NiLuJe** [comment]
Specifically, one potentially confusing factor is that the first boot after the forced poweroff might take an extra long time because of filesystem corruption, and that can take a non-trivial amount of time on some devices, and this happens *before* anything shows up on screen (and possibly before the power LED starts doing the blinky dance).

(e.g., it can take north of 45s on >= 32GB devices).

**Comment by chris-montero** [comment]
> I happen to own a day one Forma, so I'm intimately familiar with what happens when the driver decides to commit seppuku.

That's ok and I believe you, but that's not what I meant. When I said you're wrong I meant to correct you in that I did not immediately try to do a factory reset, or that I accidentally performed one. I wanted to clarify that I did try other things before resorting to that.

If something like this happens again, do you have any advice for what I could do to save the device (and more importantly the content on it) before resorting to a factory reset?

**Comment by NiLuJe** [comment]
Assuming it really refuses to power on and wasn't just still busy doing a fsck, not really.

**Comment by NiLuJe** [comment]
Oh, leaving it plugged in to a wall charger for a while might also help. (It'll trip a power on all on its own on most devices, Forma included).

**Comment by chris-montero** [comment]
@NiLuJe I see, thanks for the info and the support. You guys can do what you want with this issue. I'll just open another one if something like this happens again. Thank you for your work on this software, and have a nice day!

**Comment by Huahddd** [comment]
> > I will try to get some crash logs if I can如果可以的话，我会尝试获取一些崩溃日志
> 
> FWIW, unless you can catch it via serial, you likely won't, as these hangs do not happen while Wi-Fi is on, or while plugged in, which excludes monitoring over SSH either way.FWIW，除非您可以通过串行捕获它，否则您很可能不会，因为这些挂起不会在 Wi-Fi 开启或插入时发生，这不包括通过 SSH 进行的任何一种方式的监控。
> 
> And since the kernel hangs, nothing will get flushed to disk, so you're unlikely to see anything relevant anywhere persist on reboot.由于内核挂起，不会有任何内容刷新到磁盘，因此在重新启动时，您不太可能在任何地方看到任何相关内容。
> 
> (Again, lotsa experience here, I've tried all kinds of things to try to catch a trace of these on that specific device, to no avail).（同样，这里有很多经验，我尝试了各种方法，试图在那个特定的设备上捕捉到这些的痕迹，但无济于事）。

I want to know if the random hangs of forma have been fixed. I need to turn on wifi to ensure that forma does not randomly hang, which is very battery consuming and annoying. Also, koreader's wifi often turns off automatically (when I turn off the screen) even if I turn on the "auto reconnect to wifi" option. I want to know if there is any way to prevent forma from randomly hanging besides connecting to wifi and charging.

**Comment by NiLuJe** [comment]
It's a hardware issue, and we have no software workarounds available to us on those devices.
